TACC/Core-Portal
# task/FP-391: Notifications

## Overview: ##
Enables Django Channels on the backend, running as a daphne server for websocket events.

Creates a `ReconnectingWebSocket` connection on client instantiation. The channels ws connection is added to two groups:
- the user's `username` group, for messages to the user
- the `portal_events` group, for messages to all users

Credit to @jarosenb for laying the groundwork.

## PR Status: ##


## Related Jira tickets: ##

* [FP-391](https://jira.tacc.utexas.edu/browse/FP-391)
* [FP-81](https://jira.tacc.utexas.edu/browse/FP-81)
* [FP-27](https://jira.tacc.utexas.edu/browse/FP-27)

## Summary of Changes: ##

## Testing Steps: ##
1. Install updated `requirements.txt` by rebuilding image: 
```
make build
```
2. Install updated `packages.json`: 
```
cd client && npm ci
npm run dev
```
3. Run an [ngrok](https://ngrok.com/download) session to route webhooks to `frontera_prtl_nginx`:
```
ngrok http 443
```
4. Take url generated by ngrok and paste it into the `_WH_BASE_URL` setting in `settings_secret.py`
5. Launch a job and see the History notifications badge increment as the job status changes.


## UI Photos:
<img width="814" alt="Screen Shot 2020-07-02 at 4 20 44 PM" src="https://user-images.githubusercontent.com/20326896/86410286-21600c80-bc80-11ea-9ec9-4d09408d0801.png">

## Notes: ##
- Notification toasts and tables will be coming in a separate PR.
- There is a companion PR for the camino side changes to the websocket docker container at https://github.com/TACC/Camino/pull/3

## Todo: ##
- [ ]  If 3 job status updates come through for one job, that increments the badge counter by 1, not 3
